### ISSUE TITLE
RiichiLab ranked自分対局をtraining / analysis向けdurable raw recordとして保存する

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-arena#168]
# Summary

RiichiLab rankedでlisjong自身が参加した1半荘について、現在のopt-in protocol traceを再利用しつつ、process終了後もtraining / offline analysisのraw sourceとしてstrictにreadback可能な **durable ranked game record v0** を保存できるようにする。

現在Arenaには既に:

```text
RiichiLab ranked connection
    ↓
RankedSession
    ↓
request_action
    ↓
RiichiLabSeatAdapter
    ↓
Policy
    ↓
send action

+ opt-in JsonlProtocolTraceWriter
```

があり、`--trace` / `--trace-path` によりprotocol eventをJSONLへ保存できる。

したがって本Issueではprotocol loggingを再実装しない。

既存traceをsource of truthとして、

```text
successful RiichiLab ranked hanchan
        ↓
raw protocol trace
        +
completion / provenance manifest
        +
RankedGameResult semantics
        ↓
validated durable bundle
        ↓
process exit
        ↓
strict loader
        ↓
future dataset builder / offline analysis
```

を成立させる。

---

# Motivation

Tenhou等の外部human game logの利用可否とは独立して、lisjong自身がRiichiLabで実際に対局した際にbotへ提示されたplayer-visible informationと、lisjong自身が返したactionを研究用raw dataとして継続保存できるようにしたい。

主な将来consumerは:

- Learned Policy training source
- offline policy analysis
- failure diagnosis
- action-distribution analysis
- HandBelief input-distribution analysis
- local self-playとexternal AI-vs-AI distributionの比較

である。

ただし本Issueではtraining datasetそのものは作らない。

```text
RiichiLab durable raw record
        ↓
future consumer-specific dataset builder
        ↓
features / labels / experiments
```

の境界を維持する。

---

# Existing implementation to KEEP

implementation開始時にcurrent mainを再確認するが、少なくとも現在の以下のcontractをreuseする。

```text
lisjong_arena.riichilab.ranked
    run_ranked_game()
    RankedGameResult

lisjong_arena.riichilab.session
    RankedSession
    request_id lifecycle
    action_ack lifecycle

lisjong_arena.riichilab.transport
    drive_ranked_session()

lisjong_arena.riichilab.trace
    JsonlProtocolTraceWriter

lisjong_arena.riichilab.request_action
    parse_request_action()

lisjong_arena.riichilab.adapter
    RiichiLabSeatAdapter
```

現在のprotocol traceが既に記録する:

```text
recv protocol events
send protocol actions
timestamp
direction
event_type
raw JSON payload
```

をduplicate implementationしない。

credential / Authorization headerをtraceへ流さないcurrent secret-safe boundaryも維持する。

---

# Scope

## Included

対象execution path:

```text
RiichiLab ranked
1 connection
1 completed hanchan
lisjong-controlled Policy
```

保存対象:

```text
raw received protocol events
raw attempted outgoing action payloads
ranked completion/result metadata
record provenance
integrity metadata
```

storage:

```text
operator-declared local non-ephemeral filesystem
```

readback:

```text
cross-process strict loader
```

primary consumer:

```text
future training / offline analysis source
```

## Deferred

- RiichiLab公開Game一覧のscraping
- RiichiLab server-side full game log download
- 他bot・他ユーザーのbulk log acquisition
- Tenhou log ingestion
- Mortal log ingestion
- S3 / AWS upload
- database
- automatic continuous queue / daemon
- model training
- dataset split
- HandBelief labels
- hidden-hand ground truth
- viewer / replay UI

---

# Important data boundary

本recordは **lisjong botが実際に観測・送受信した情報** を保存する。

特に`request_action`についてraw payloadをlosslessに保持し、少なくとも:

```text
request_id
possible_actions
observation
time metadata if present
unknown future-compatible fields if present
```

を落とさない。

`observation`はserverから受け取ったbase64 representationをraw sourceとして保持する。

readback時にはexisting `parse_request_action()` または同じcanonical RiichiEnv deserialization pathを利用して、Observationとして復元可能であることを確認する。

---

# Player-visible vs omniscient truth

このrecordをomniscient game recordとして扱わない。

```text
RiichiLab request_action Observation
    = serving時にlisjong自身が見えていたplayer-safe state
```

である。

したがって:

```text
opponent concealed hand ground truth
wall ground truth
future draw truth
```

等を本recordが持つことを要求しない。

特にHandBelief用途では:

```text
usable:
    model input / online distribution
    lisjong自身のdecision context

not guaranteed:
    opponent hidden-hand teacher label
```

という境界を明記する。

---

# Minimum durable record bundle

physical file namingはcurrent repository conventionsをpreflightして決定してよい。

logical bundleとして最低限:

```text
manifest
protocol trace
final ranked result
```

を保持する。

conceptual example:

```text
<record>/
    manifest.json
    protocol.jsonl
    result.json
```

exact physical layoutはcontractではなく、logical semanticsを優先する。

---

# Manifest / provenance

最低限:

```text
record schema id
record schema version

mode = ranked
completion status

bound seat

profile identity
Policy identity where resolvable

Arena source revision
lisjong dependency/source revision where resolvable
RiichiEnv version
Python/runtime information where useful

protocol trace digest
result digest

record identity
```

を保持する。

revision / Policy identity等を完全にresolveできない環境では、推測値で埋めず `unresolved` 相当として明示する。

以下はsemantic record identityへ含めない。

```text
absolute local path
machine username
timestamp alone
mutable storage location
```

---

# Raw protocol trace

current `JsonlProtocolTraceWriter` semanticsを可能な限りreuseする。

受信eventについて:

```text
direction = recv
event_type
payload
```

送信actionについて:

```text
direction = send
event_type
payload
```

を保持する。

現在の意味どおり、send traceはactual server acceptanceではなく、`transport.send()`を試みる直前に生成されたsend-ready actionを意味する。

actual acceptanceについては後続`action_ack`を別protocol eventとして解釈する。

---

# Decision correlation

strict readbackでは少なくとも`request_id`を用いて:

```text
request_action
      ↓
outgoing action
      ↓
action_ack(s)
```

をcorrelateできること。

completed recordについて最低限:

- request IDs are valid
- request IDs follow current RankedSession lifecycle semantics
- each sent response corresponds to a known request
- outgoing payload request_id matches request
- recorded request / response counts are consistent with RankedGameResult
- ack references only known requests
- fatal ack semanticsをsuccessful completed recordとして偽装しない

を検証する。

全requestについて必ず特定個数のackが存在する、という新しいprotocol assumptionは追加しない。

---

# Small training / analysis readback smoke

training dataset implementationは行わない。

ただしraw recordのutilityを確認するため、loaderから少なくともconceptually:

```text
for each request_action:
    request_id
    bound seat
    deserialized Observation
    possible_actions
    sent action
    ack history
```

を取得できるsmall deterministic readback pathを提供する。

このreadback pathでObservationからexisting Arena / lisjong projectionが可能であることは確認してよい。

ただし本Issueで:

```text
PolicyInput tensor
action vocabulary index
reward
Q target
BC label
HandBelief label
```

へ変換しない。

---

# Completion semantics

current traceはdiagnostic用途として途中まで残る場合がある。

その性質は維持してよい。

ただし:

```text
partial diagnostic trace
!=
completed durable ranked record
```

とする。

successful durable bundleは:

```text
ranked execution reaches valid end_game
        ↓
RankedGameResult obtained
        ↓
trace closed successfully
        ↓
manifest / result serialize
        ↓
digests computed
        ↓
strict readback validation
        ↓
atomic / bounded finalization
```

後にのみcompletedとして公開する。

---

# Failure semantics

以下をvalid completed recordとして扱わない。

- connection failure
- unexpected disconnect before `end_game`
- Policy / Adapter failure
- malformed protocol event
- fatal `action_ack`
- trace write / close failure
- result serialization failure
- digest mismatch
- strict readback failure
- incomplete temporary bundle

Diagnostic目的のpartial traceを残すこと自体は禁止しない。

ただしloaderがcompleted durable recordとしてacceptしてはならない。

---

# Integrity / identity

minimum integrityとして各semantic payloadへdigestを持つ。

```text
protocol trace digest
result digest
manifest semantic fields
```

からrecord identityをdeterministically検証可能にする。

timestamp / UUIDだけをrecord identityとして使用しない。

same Policy / same seatであっても、protocol payloadが異なるgameは異なるrecord identityになる。

---

# Existing target / append semantics

current `JsonlProtocolTraceWriter`は指定fileへappendできるが、durable ranked recordでは複数gameを暗黙に1 recordへ混在させない。

```text
1 durable record
=
1 RiichiLab ranked hanchan
```

とする。

existing completed recordへのsilent overwrite / silent appendを行わない。

---

# Secret / privacy safety

current secret-safe contractを維持する。

recordへ以下を保存しない。

```text
BOT_TOKEN
Authorization header
credential environment value
other authentication material
```

credentialを一度書き込んでからredactする方式にしない。

また本Issueは、自分のbot sessionで受信した情報をlocal recordとして取得する機能に限定する。

このIssue自体は:

- third-party game scraping
- data redistribution permission
- public dataset publishing rights

について新しい権利判断を行わない。

---

# Operator-facing acquisition

existing ranked CLIへ薄くcompositionする。

conceptual example:

```text
python -m lisjong_arena.riichilab.ranked \
    --profile lisjong \
    --record-dir <path>
```

exact CLI namingはcurrent `--trace` contractとの混同を避けつつpreflightで決定する。

重要なのはdiagnostic traceとcompleted durable research recordを意味的に混同しないことである。

既存`--trace` compatibilityを壊さない。

---

# Relationship to Arena #155

Arena #155 `Training / analysis向けdurable local game record` はstandard RiichiEnv `LocalGameRunner`を対象としており、RiichiLabは明示的にdeferredされている。

本Issueはそのscopeを変更せず、RiichiLab external participation側のconcrete consumerとして独立させる。

```text
#155
    local RiichiEnv durable record

this Issue
    RiichiLab ranked durable record
```

とする。

両者を今すぐproject-wide canonical `GameRecord`へ統合しない。

implementation開始時に#155が既にreusableなdigest helper / atomic finalization helper / strict manifest helper等を提供している場合は、意味境界を壊さない範囲でthin reuseしてよい。

ただし#155完了を本Issueの必須blockerにはしない。

---

# Ownership

Primary owner:

```text
lisbun/lisjong-arena
```

理由:

- RiichiLab ranked executionはArena-owned
- Session / Transport / protocol traceはArena-owned
- game acquisition / observationはArena責務
- downstream training / evaluationとの接続もArena側が自然
- stable lisjong Policy contractを変更する必要がない

`lisjong`へRiichiLab persistence責務を逆流させない。

---

# Preflight

implementation開始時に最低限確認する。

- this Issue / latest comments
- Arena #23
- Arena #27
- Arena #155 current state
- current `AGENTS.md`
- `README.md`
- `docs/architecture.md`
- `docs/riichilab-client.md`
- `docs/riichilab-protocol-bridge.md`
- `src/lisjong_arena/riichilab/ranked.py`
- `src/lisjong_arena/riichilab/session.py`
- `src/lisjong_arena/riichilab/transport.py`
- `src/lisjong_arena/riichilab/trace.py`
- `src/lisjong_arena/riichilab/request_action.py`
- `src/lisjong_arena/riichilab/adapter.py`
- current profile / provenance helpers
- current artifact / digest / atomic-write helpers where appropriate
- current RiichiLab protocol documentation

---

# Tests

## Record round-trip

- valid completed ranked record round-trip
- schema/version validation
- trace digest validation
- result digest validation
- record identity re-derivation
- different storage path does not alter semantic identity

## Protocol preservation

- `request_action` raw payload preserved
- base64 Observation can be deserialized through canonical parser
- `possible_actions` preserved
- unknown extra fields preserved
- sent action payload preserved
- request_id correlation preserved
- action_ack history preserved

## Completion

- valid `end_game` required for completed record
- request/response counts consistent
- bound seat consistent
- result metadata consistent

## Failure

- truncated trace rejected
- corrupt JSON rejected
- digest mismatch rejected
- unsupported schema rejected
- incomplete bundle rejected
- unexpected disconnect not finalized
- trace write failure not finalized
- fatal protocol failure not finalized
- existing completed target not silently overwritten/appended

## Secret safety

- token not persisted
- Authorization header not persisted
- credential environment value not persisted

## Integration

real RiichiLab network accessをCI requirementにしない。

existing fake Transport / Session fixturesをreuseして:

```text
synthetic ranked session
    ↓
record
    ↓
finalize
    ↓
new loader
    ↓
strict readback
```

を検証する。

manual real-ranked smokeを行う場合も、CI successの必須条件にはしない。

---

# Documentation

最低限以下を同期する。

- `README.md`
- `docs/architecture.md`
- `docs/riichilab-client.md`
- purpose-specific durable record document if justified

次を明記する。

```text
protocol trace
    diagnostic / wire-level evidence

RiichiLab durable ranked record
    completed one-hanchan research raw source

training dataset
    downstream consumer-specific artifact

omniscient game log
    not provided
```

---

# Non-goals

- project-wide canonical `GameRecord`
- local Arena #155 schemaとの強制統一
- public RiichiLab game scraping
- bulk download API implementation
- other users' game collection
- server-side omniscient game log acquisition
- hidden opponent hand reconstruction
- HandBelief ground-truth labels
- training tensor schema
- BC dataset
- RL transitions / rewards
- model training
- Champion promotion
- ranked requeue / continuous daemon
- retry / reconnect framework
- S3 / cloud storage
- database / registry
- automatic retention policy
- replay GUI
- Tenhou ingestion
- Mortal ingestion
- credential persistence

---

# Acceptance criteria

- [ ] current RiichiLab trace implementationをreuseし、protocol loggerを再実装していない
- [ ] scopeを自分のbotが参加するRiichiLab ranked 1半荘へ限定した
- [ ] 1 completed hanchan = 1 durable recordになっている
- [ ] raw recv/send protocol payloadを保持する
- [ ] `request_action.observation` raw representationを保持する
- [ ] existing parserでObservationをreadbackできる
- [ ] `possible_actions`を保持する
- [ ] sent actionを保持する
- [ ] request / response / ackをrequest_idでcorrelateできる
- [ ] `RankedGameResult` semanticsをrecordへbindする
- [ ] schema id / versionを持つ
- [ ] provenanceを可能な範囲で明示する
- [ ] payload digest / record identityをstrictに再検証できる
- [ ] partial / corrupt / tampered recordをfail closedで拒否する
- [ ] completed targetへのsilent overwrite / appendを行わない
- [ ] BOT_TOKEN / Authorization等credentialを保存しない
- [ ] player-visible recordとomniscient truthを混同していない
- [ ] small readback smokeでfuture dataset sourceとしてのutilityを確認した
- [ ] tensor / label / trainingを本Issueへ含めていない
- [ ] public game scraping / bulk acquisitionへscopeを広げていない
- [ ] Arena #155をreopen / scope expansionしていない
- [ ] project-wide canonical GameRecordを導入していない
- [ ] real RiichiLab network accessをCI requirementにしていない
- [ ] affected documentationを同期した

**Comment by lisbun** [comment]
## 2026-09-12 status refresh

The Issue remains valid, but its relationship to Arena #155 has advanced.

```text
#155  COMPLETE — durable local game record / strict loader
#207  COMPLETE — schema v2 authoritative per-round result facts
#168  PLANNED — RiichiLab-ranked-specific durable raw record
```

Therefore implementation should now preflight the completed durable-record implementation and **reuse its proven integrity patterns where semantically compatible** (versioning, digest/identity validation, atomic finalize, strict fail-closed readback), rather than treating #155 as merely a possible future helper source.

The architecture boundary remains unchanged:

```text
RiichiEnv durable local record schema
!= RiichiLab durable ranked record schema
```

Do not force the two execution paths into one canonical record merely because common persistence helpers now exist. #168 should still start from RiichiLab's concrete acquisition/consumer requirements and keep credential/session data out of the durable research record.

**Comment by lisbun** [comment]
## Real RiichiLab ranked durable-record smoke — PASS

#168で実装したdurable ranked recordを、`lisjong-dev` / `MechanismRiichiDefenseYakuhaiCallPolicy` の実RiichiLab ranked 1半荘でoperator smokeした。

Result:

```text
completed hanchan: yes
bound seat: 1
requests: 132
responses: 132
acknowledged requests: 132
deserialized observations: 132
final scores: (2400, 41000, 21200, 35400)
record identity:
97f20dadcbdccedd6699a38947f6d30b3b479fbaf5c3d5e9dd2b07f56c39470c
```

Strict readback succeeded with:

```text
execution_environment = riichilab-ranked
profile_identity      = lisjong-dev
policy_identity       = MechanismRiichiDefenseYakuhaiCallPolicy
RiichiEnv              = 0.4.10
Python                 = CPython 3.14.6
Arena revision         = f0259f082abc45ed31c317044d50dfcc663e70c0
lisjong revision       = d1d3c14e3e1948cff871273b7b8b57a08c655db9
lisjong-engine revision= 8735e89e1aea000ab59368d0368d476787827741
```

The persisted bundle contains the expected completed-record payloads:

```text
manifest.json
protocol.jsonl  550,399 bytes
result.json      11,373 bytes
```

The payload SHA-256 values recorded in `manifest.json` matched the uploaded/local files, and the record identity matched the identity printed by the ranked acquisition command.

Conclusion:

**PASS — #168 durable ranked record has now been exercised successfully on a real completed RiichiLab ranked hanchan, including publish, integrity validation, canonical Observation readback, request/response/ack correlation, and provenance readback.**

This is persistence / compatibility evidence only. The game result itself (seat 1, 41,000 points) is not treated as strength evidence.
